src/services/identity/UsersService.ts

```diff
@@ -208,201 +212,169 @@ class UsersService {
+    //  time with AsyncResult.combine() wrappers (which makes the code har to read)
+    let transaction: Transaction | null = null
+    let otpEntry: Otp | null = null
+
+    // TypeScript can't tell when transaction and otpEntry are guaranteed to not be null in the promise chain steps
+    // So we'll provide non-null assertions ourselves, and we need to tell esling to leave us alone -_-
+    /* eslint-disable @typescript-eslint/no-non-null-assertion */
+    return ResultAsync.fromPromise(this.sequelize.transaction(), (error) =>
+      this.otpGetAndLogError(
+        DatabaseError,
+        error,
+        "Error starting database transaction"
+      )
+    )
+      .andThen((_transaction) => {
+        transaction = _transaction // store transaction in outer scope, so it is accessible all the way to the orElse() below
+
+        return ResultAsync.fromPromise(
+          this.otpRepository.findOne({
+            where: findConditions,
+            lock: true,
+            transaction,
+          }),
+          (error) =>
+            this.otpGetAndLogError(DatabaseError, error, findErrorMessage)
+        )
       })
-      .andThen(() => {
+      .andThen((_otpEntry: Otp | null) => {
+        otpEntry = _otpEntry // store otpEntry in outer scope, so it's easier to access it in the local promise chain steps
+
+        // verify that otpDbEntry exists
         if (!otpEntry) {
           return errAsync(new BadRequestError("OTP not found"))
         }
 
-        return okAsync(otpEntry)
-      })
-      .andThen((otpDbEntry) => {
-        if (otpDbEntry.attempts >= MAX_NUM_OTP_ATTEMPTS) {
+        // after this point, otpEntry is guaranteed to be truthy is all promise chain steps (just like transaction)
+
+        // verify otpEntry validity
+
+        if (otpEntry.expiresAt < new Date()) {
+          return this.otpDestroyEntry(otpEntry, transaction!).andThen(() =>
+            errAsync(new BadRequestError("OTP has expired"))
+          )
+        }
+
+        if (otpEntry.attempts >= MAX_NUM_OTP_ATTEMPTS) {
+          // should this delete the otpEntry as well?
           return errAsync(new BadRequestError("Max number of attempts reached"))
         }
 
-        return okAsync(otpDbEntry)
-      })
-      .andThen((otpDbEntry) => {
-        if (!otpDbEntry.hashedOtp) {
-          return ResultAsync.fromPromise(
-            otpDbEntry.destroy({ transaction }),
-            (error) => {
-              logger.error(
-                `Error destroying OTP entry: ${JSON.stringify(error)}`
-              )
-
-              return new DatabaseError("Error destroying OTP entry in database")
-            }
-          ).andThen(() => errAsync(new BadRequestError("Hashed OTP not found")))
+        if (!otpEntry.hashedOtp) {
+          return this.otpDestroyEntry(otpEntry, transaction!).andThen(() =>
+            errAsync(new BadRequestError("Hashed OTP not found"))
+          )
         }
```

Comment:
This should probably be done in the DB query itself where we we only want record where `hashedOtp` is a truthy value.

Reply:
Possible implementation for the query:
```
            where: {
              ...findConditions,
              hashedOtp: {
                [Op.regexp]: '\\S+' // at least one non-white character
              }
            },
```

Reply:
done